datafusion/expr/src/columnar_value.rs

```diff
@@ -195,19 +195,39 @@ impl ColumnarValue {
                 kernels::cast::cast_with_options(array, cast_type, &cast_options)?,
             )),
             ColumnarValue::Scalar(scalar) => {
-                let scalar_array =
-                    if cast_type == &DataType::Timestamp(TimeUnit::Nanosecond, None) {
-                        if let ScalarValue::Float64(Some(float_ts)) = scalar {
-                            ScalarValue::Int64(Some(
-                                (float_ts * 1_000_000_000_f64).trunc() as i64,
-                            ))
-                            .to_array()?
-                        } else {
-                            scalar.to_array()?
-                        }
+                let scalar_array = if cast_type
+                    == &DataType::Timestamp(TimeUnit::Nanosecond, None)
+                {
+                    if let ScalarValue::Float64(Some(float_ts)) = scalar {
+                        ScalarValue::Int64(Some(
+                            (float_ts * 1_000_000_000_f64).trunc() as i64
+                        ))
+                        .to_array()?
                     } else {
                         scalar.to_array()?
-                    };
+                    }
+                } else {
+                    // Arrow by default will parse str as Month for unit MonthDayNano.
+                    // So we need to be explict that we want it to parse as second.
+                    match (scalar, cast_type) {
```

Comment:
one issue with this approach is that it will only work for single values like `'1'::interval` 

the behavior of 
```sql
create table foo as values ('1');
select column1::interval from foo;
```

will not be affected

I recommend you update the code in `kernels::cast` to have the new semantics you have in mind

Reply:
Sorry but as my understand is that `kernels::cast` is within `arrow-cast` which it a different repo. So you mean this change is should happen in `arrow-cast` repo ?

Reply:
Yes, eventally the fix should make it up to arrow-cast

In the interim, you could make a wrapper function in datafusion with the same signature as `cast` in arrow and implement your logic there for interval and fall through to the kernel in arrow-cast otherwise

I haven't though through the implications of changing the semantics of casting strings to intervals to be honest